test/test_torch.py

```diff
@@ -14965,6 +14965,19 @@ def test_cat(self, device, dtype):
         self.assertRaises(RuntimeError, lambda: torch.cat([]))
         self.assertRaisesRegex(TypeError, 'got None', lambda: torch.cat([x, None]))
 
+    @onlyCPU
+    def test_cat_different_dtypes(self, device):
+        x = torch.tensor([1, 2, 3], device=device, dtype=torch.int8)
+        y = torch.tensor([4, 5, 6], device=device, dtype=torch.int32)
+        expected_out = torch.tensor([1, 2, 3, 4, 5, 6], device=device, dtype=torch.int32)
+        out = torch.cat([x, y])
+        self.assertEqual(out, expected_out) 
+        z = torch.tensor([7, 8, 9], device=device, dtype=torch.int16)
+        expected_out = torch.tensor([1, 2, 3, 4, 5, 6, 7, 8, 9],
+                                    device=device, dtype=torch.int32)
+        out = torch.cat([x, y, z])
```

Comment:
is there also a torch.cat(tensors, out=result) variant we should check? In most functions with out variants we validate the out dtype, e.g. you can perform ops like `torch.add(int_tensor, double_tensor, out=float_tensor)` but not `torch.add(int_tensor, double_tensor, out=int_tensor)`. When we do validate `out=` variants we rely on the `canCast` function

Reply:
When using cat with `out` parameter, An error check is made and raised if the `dtypes` of the Tensors are different from the `out`.

Should we cast the inputs here too? It will be more consistent.

Reply:
I'm thinking it'd be something like:
```
assert( canCast( high_type, out.dtype ));
// calculate result
out.copy_(result);
return out;
```
it's not really a cast because an `out` tensor has to be modified in-place. The dtypes of ten tensors should be able to differ from `out` but the canCast check defines the specific cases where they can. E.g. cat( (ints, doubles), out=floats) should be allowed but not `cat( (ints, floats), out=ints)`.

Reply:
I took another approach to solve this using the `TensorIterator` class to automatically cast the type to the output tensor, this works in both the `cat` with out and without it.

Reply:
this looks simpler and I like the consistency with other implementations too. It'll still be good to add tests for any interesting cases like `out=` though